### DISCUSSION TITLE
Emgu and App SDK ( WinUI 3 )

### DISCUSSION THREAD

**Discussion by onslauth** [emgucv/emgucv#730] [Q&A]
Hi,

I am new to Emgu, and I am currently evaluating it for use in a project where I need to take existing python code using opencv and tensorflow and convert it to C#, specifically for use with App SDK ( WinUI 3 ).

The code below converts from a `mat` object to a `image source` that can be used by the `<Image x:Name="MainImage" />` that is added to the XAML.
What I would like to know is if there is a better way to achieve the following? 

```C#
            CvInvoke.Init( );
            string sentence = "Testing EMGU OpenCV";
            Mat img = new Mat( 400, 400, Emgu.CV.CvEnum.DepthType.Cv8U, 3 );
            img.SetTo( new Bgr( 255, 0, 0 ).MCvScalar );

            CvInvoke.PutText( img, sentence, new System.Drawing.Point( 10, 80 ),
                Emgu.CV.CvEnum.FontFace.HersheyComplex, 1.0, new Bgr( 0, 255, 0 ).MCvScalar );

            Image< Bgr, Byte > image = img.ToImage< Bgr, Byte >( );
            var stream = image.ToJpegData( ).AsBuffer( ).AsStream( ).AsRandomAccessStream( );
            BitmapDecoder decoder = await BitmapDecoder.CreateAsync( stream );
            SoftwareBitmap sbmp = await decoder.GetSoftwareBitmapAsync( );

            if ( sbmp.BitmapPixelFormat != BitmapPixelFormat.Bgra8 || sbmp.BitmapAlphaMode == BitmapAlphaMode.Straight )
            {
                sbmp = SoftwareBitmap.Convert( sbmp, BitmapPixelFormat.Bgra8, BitmapAlphaMode.Premultiplied );
            }

            SoftwareBitmapSource sbmps = new SoftwareBitmapSource( );
            await sbmps.SetBitmapAsync( sbmp );
            this.MainImage.Source = sbmps;
```

**Comment by emgucv** [comment]
You can use this to convert IInputArray (including Image<,>) to BitmapSource:
https://github.com/emgucv/emgucv/blob/master/Emgu.CV.NativeImage/BitmapSourceExtension.cs

The above code directly copy the bitmap data without running a JPEG compression. It should be more efficient.

**Reply by onslauth** [reply]
Thank you very much for the above code. I have added it to my solution.

**Reply by chungheecha** [reply]
It is not easy to add the extension.
Would you kindly let mw know how to add successfully ?

**Reply by emgucv** [reply]
For the 4.6.0 release, you can also add the Emgu.CV.Wpf nuget package:
https://www.nuget.org/packages/Emgu.CV.Wpf

**Reply by chungheecha** [reply]
Thanks.
Still do not know.....

**Reply by deepakgh** [reply]
https://github.com/emgucv/emgucv/blob/master/Emgu.CV.NativeImage/BitmapSourceExtension.cs 
is dependent on PresentationCore.dll 
Is there any way to use it in WinUI?
What other option is there to create a BitmapSpurce in WinUI?
